<a href="https://colab.research.google.com/github/LitvinovaYaI/ML_Credit_cards/blob/main/ML_task2_part2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [15]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.svm import LinearSVC
from sklearn.metrics import classification_report
from sklearn.metrics import roc_curve
from sklearn.metrics import roc_auc_score
from sklearn.ensemble import RandomForestClassifier
from keras.models import Sequential
from keras.layers import InputLayer, Dense, BatchNormalization
from keras.callbacks import ModelCheckpoint

In [16]:
data = pd.read_csv("/content/creditcard.csv")
data.head()
data.info()
data = data.drop_duplicates(keep='first')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [17]:
impostor = data[data['Class']==1]
legal = data[data['Class']==0]
legal_sampled = data[data['Class']==0].sample(n = impostor.shape[0], random_state=21)
print(f' Impostor Shape{impostor.shape}\n Legal shape{legal_sampled.shape}')
impostor_sampled = impostor
final_data = pd.concat([impostor_sampled, legal_sampled], ignore_index=True)
final_data.shape

 Impostor Shape(473, 31)
 Legal shape(473, 31)


(946, 31)

In [18]:
x_dropped=final_data.drop(data.columns[[0,1,2,3,4,5,6,7,8,9,10,11,13,15,16,20,21,22,23,24,25,26,27,28,29,30]], axis=1)
print(x_dropped)
scaler=StandardScaler()
x_dropped = scaler.fit_transform(x_dropped)
y=final_data['Class']
x_dropped, x_dropped_test, y_dropped, y_dropped_test = train_test_split(x_dropped,y, test_size=0.20, random_state=21)
print(f'x_train{x_dropped.shape}\n, x_test{x_dropped_test.shape}\n, y_train{y_dropped.shape}\n, y_test{y_dropped_test.shape}')

           V12       V14        V17       V18       V19
0    -2.899907 -4.289254  -2.830056 -0.016822  0.416956
1    -0.503141 -1.692029   0.599717  1.725321  0.283345
2    -6.560124 -1.470102  -4.781831 -2.615665 -1.334441
3   -10.912819 -6.771097 -12.598419 -5.131549  0.308334
4    -4.609628 -6.079337   6.739384  3.042493 -2.721853
..         ...       ...        ...       ...       ...
941   0.137928  0.592303  -1.100491  0.736335 -0.327121
942   1.604825  0.292753  -0.498097 -0.317072 -0.534416
943   0.314660 -0.035399   0.236162  0.230408 -1.244823
944   0.714843 -1.111186   0.332380  0.026919  0.253231
945   0.198548  0.800798  -0.145792  0.352731  1.148725

[946 rows x 5 columns]
x_train(756, 5)
, x_test(190, 5)
, y_train(756,)
, y_test(190,)


In [19]:
# Logistic Regression model
def logReg(train_x, train_y, x_test, y_test):
    logistic_model = LogisticRegression()
    logistic_model.fit(train_x, train_y)
    y_pred_train = logistic_model.predict(train_x)
    y_pred_test = logistic_model.predict(x_test)
    res_train = classification_report(train_y, y_pred_train, target_names=["Imposter","Legal"])
    res_test = classification_report(y_test, y_pred_test, target_names=["Imposter","Legal"])
    print(res_train)
    print("*"*100)
    print(res_test)

In [20]:
#KNeighbors classification model
def KNeighbors(x_train, y_train, x_test, y_test):
  Kneib = KNeighborsClassifier(n_neighbors=3)
  Kneib.fit(x_train, y_train)
  y_pred_train = Kneib.predict(x_train)
  y_pred_test = Kneib.predict(x_test)
  res_train = classification_report(y_train, y_pred_train, target_names=["Imposter","Legal"])
  res_test = classification_report(y_test, y_pred_test, target_names=["Imposter","Legal"])
  print(res_train)
  print("*"*100)
  print(res_test)

In [21]:
LG_dropped = logReg(x_dropped, y_dropped, x_dropped_test, y_dropped_test)

              precision    recall  f1-score   support

    Imposter       0.89      0.98      0.93       380
       Legal       0.97      0.88      0.92       376

    accuracy                           0.93       756
   macro avg       0.93      0.93      0.93       756
weighted avg       0.93      0.93      0.93       756

****************************************************************************************************
              precision    recall  f1-score   support

    Imposter       0.83      0.99      0.90        93
       Legal       0.99      0.80      0.89        97

    accuracy                           0.89       190
   macro avg       0.91      0.90      0.89       190
weighted avg       0.91      0.89      0.89       190



In [22]:
KN_dropped = KNeighbors(x_dropped, y_dropped, x_dropped_test, y_dropped_test)

              precision    recall  f1-score   support

    Imposter       0.91      0.98      0.95       380
       Legal       0.98      0.90      0.94       376

    accuracy                           0.94       756
   macro avg       0.95      0.94      0.94       756
weighted avg       0.95      0.94      0.94       756

****************************************************************************************************
              precision    recall  f1-score   support

    Imposter       0.85      0.97      0.90        93
       Legal       0.96      0.84      0.90        97

    accuracy                           0.90       190
   macro avg       0.91      0.90      0.90       190
weighted avg       0.91      0.90      0.90       190



Итак, результаты для полного датасета: метод линейной регрессии 94%, метод ближайших соседей 97%, для обрезанного датасета: метод линейной регрессии 91%, метод ближайших соседей 91%. Сохранение только наиболее коррелируемых параметров не улучшило результат.